dissect/target/containers/fortifw.py

```diff
@@ -0,0 +1,206 @@
+        logger.info("gzipped: %s", self.is_gzipped)
+        self.fh.seek(0)
+
+    def seek(self, offset, whence=io.SEEK_SET):
+        return self.fh.seek(offset, whence)
+
+    def read(self, n=-1):
+        data = bytearray()
+
+        while True:
+            pos = self.fh.tell()
+            buf = self.fh.read(io.DEFAULT_BUFFER_SIZE)
+            if not buf:
+                break
+
+            if self.xor_key:
+                if pos % 512 == 0:
+                    xor_char = 0xFF
+                else:
+                    self.fh.seek(pos - 1)
+                    xor_char = ord(self.fh.read(1))
+                    self.fh.seek(pos + len(buf))
+
+                for i, cur_char in enumerate(buf):
+                    if (i + pos) % 512 == 0:
+                        xor_char = 0xFF
+                    idx = (i + pos) & 0x1F
+                    data.append(((self.xor_key[idx] ^ cur_char ^ xor_char) - idx) & 0xFF)
+                    xor_char = cur_char
+            else:
+                data.extend(buf)
+
+            if n > 0 and len(data) >= n:
+                break
+
+        if n == -1:
+            n = None
+        return bytes(data[:n])
+
+
+class FortiFirmwareContainer(Container):
+    __type__ = "fortifw"
+
+    def __init__(self, fh: Union[BinaryIO, Path], *args, **kwargs) -> None:
+        if not hasattr(fh, "read"):
+            fh = fh.open("rb")
+
+        # Open the firmware file
+        self.ff = FortiFirmwareFile(fh)
+
+        # seek to MBR
+        self.fw = RelativeStream(self.ff, 0x200)
+        super().__init__(self.fw, self.ff.size, *args, **kwargs)
+
+    @staticmethod
+    def detect_fh(fh: BinaryIO, original: Union[list, BinaryIO]) -> bool:
+        return False
+
+    @staticmethod
+    def detect_path(path: Path, original: Union[list, BinaryIO]) -> bool:
+        # all Fortinet firmware files end with `-FORTINET.out`
+        return str(path).lower().endswith("-fortinet.out")
+
+    def read(self, length: int) -> bytes:
+        return self.fw.read(length)
+
+    def seek(self, offset: int, whence: int = io.SEEK_SET) -> int:
+        return self.fw.seek(offset, whence)
+
+    def tell(self) -> int:
+        return self.fw.tell()
+
+    def close(self) -> None:
+        pass
+
+
+@catch_sigpipe
+def main(argv: Optional[Sequence[str]] = None) -> None:
+    import argparse
+    import sys
+
+    parser = argparse.ArgumentParser(description="Decompress and deobfuscate Fortinet firmware file to stdout.")
+    parser.add_argument("file", type=argparse.FileType("rb"), help="Fortinet firmware file")
+    parser.add_argument("--verbose", "-v", action="store_true", help="Verbose output")
+    args = parser.parse_args(argv)
+
+    if args.verbose:
+        logging.basicConfig(level=logging.INFO)
+
+    ff = FortiFirmwareFile(args.file)
+    while True:
+        data = ff.read(io.DEFAULT_BUFFER_SIZE)
+        if not data:
+            break
+        sys.stdout.buffer.write(data)
```

Comment:
You can maybe use `shutil.copyfileobj(ff, sys.stdout.buffer)`

Reply:
nice one, that saves some lines :)